plugins/omdtls/omdtls.c

```diff
@@ -0,0 +1,885 @@
+			ctrType_IntCtr, CTR_FLAG_RESETTABLE, &pData->ctrDtlsFail));
+		CHKiRet(statsobj.ConstructFinalize(pData->stats));
+	}
+CODE_STD_FINALIZERnewActInst
+	cnfparamvalsDestruct(pvals, &actpblk);
+ENDnewActInst
+
+
+BEGINmodExit
+CODESTARTmodExit
+	DBGPRINTF("modExit: ENTER\n");
+	statsobj.Destruct(&dtlsStats);
+	objRelease(net_ossl, LM_NET_OSSL_FILENAME);
+	objRelease(statsobj, CORE_COMPONENT);
+	objRelease(datetime, CORE_COMPONENT);
+	objRelease(glbl, CORE_COMPONENT);
+ENDmodExit
+
+NO_LEGACY_CONF_parseSelectorAct
+BEGINqueryEtryPt
+CODESTARTqueryEtryPt
+CODEqueryEtryPt_STD_OMODTX_QUERIES
+CODEqueryEtryPt_STD_OMOD8_QUERIES
+CODEqueryEtryPt_STD_CONF2_QUERIES
+CODEqueryEtryPt_STD_CONF2_setModCnf_QUERIES
+CODEqueryEtryPt_STD_CONF2_PREPRIVDROP_QUERIES
+CODEqueryEtryPt_STD_CONF2_OMOD_QUERIES
+ENDqueryEtryPt
+
+
+BEGINmodInit()
+CODESTARTmodInit
+//	uchar *pTmp;
+	DBGPRINTF("modInit: ENTER\n");
+INITLegCnfVars
+	*ipIFVersProvided = CURR_MOD_IF_VERSION;
+CODEmodInit_QueryRegCFSLineHdlr
+	/* request objects we use */
+	CHKiRet(objUse(glbl, CORE_COMPONENT));
+	CHKiRet(objUse(net_ossl, LM_NET_OSSL_FILENAME));
+	CHKiRet(objUse(datetime, CORE_COMPONENT));
+	CHKiRet(objUse(statsobj, CORE_COMPONENT));
+
+	CHKiRet(statsobj.Construct(&dtlsStats));
+	CHKiRet(statsobj.SetName(dtlsStats, (uchar *)"omdtls"));
+	CHKiRet(statsobj.SetOrigin(dtlsStats, (uchar*)"omdtls"));
+	STATSCOUNTER_INIT(ctrDtlsSubmit, mutCtrDtlsSubmit);
+	CHKiRet(statsobj.AddCounter(dtlsStats, (uchar *)"submitted",
+		ctrType_IntCtr, CTR_FLAG_RESETTABLE, &ctrDtlsSubmit));
+	STATSCOUNTER_INIT(ctrDtlsFail, mutCtrDtlsFail);
+	CHKiRet(statsobj.AddCounter(dtlsStats, (uchar *)"failures",
+		ctrType_IntCtr, CTR_FLAG_RESETTABLE, &ctrDtlsFail));
+
+	CHKiRet(statsobj.ConstructFinalize(dtlsStats));
+ENDmodInit
+
+/*
+*	Setup socket and establish virtual Connect on a SOCKET Level for DTLS
+*/
+static rsRetVal
+dtls_send(wrkrInstanceData_t *pWrkrData, const actWrkrIParams_t *__restrict__ const pParam, const int iMsg) 
+{
+	DEFiRet;
+	int iErr, sslerr;
+	size_t lenSent = -1;
+	instanceData *const pData = (instanceData *const) pWrkrData->pData;
+
+	// Get Msg String
+	const char* pszParamStr = (const char*)actParam(pParam, 1 /*pData->iNumTpls*/, iMsg, 0).param;
+	size_t tzParamStrLen = actParam(pParam, 1 /*pData->iNumTpls*/, iMsg, 0).lenStr;
+	DBGPRINTF("dtls_send[%p]: writeProton for msg %d msg:'%.*s%s'\n",
+		pWrkrData,
+		iMsg,
+		(int)(tzParamStrLen > 0 ? (tzParamStrLen > 64 ? 64 : tzParamStrLen-1) : 0),
+		pszParamStr,
+		(tzParamStrLen > 64 ? "..." : ""));
+
+	iErr = SSL_write(pWrkrData->sslClient, pszParamStr, tzParamStrLen);
+	if(iErr > 0) {
+		DBGPRINTF("dtls_send[%p]: Successfully send message '%s' with %ld bytes to %s:%s\n",
+			pWrkrData, pszParamStr, tzParamStrLen, pData->target, pData->port);
+	} else {
+		sslerr = SSL_get_error(pWrkrData->sslClient, iErr);
+		if (sslerr == SSL_ERROR_SYSCALL) {
+			dbgprintf("dtls_send[%p]: SSL_write failed with SSL_ERROR_SYSCALL(%s)"
```

Comment:
q: would it make sense to output these error messages also to the user? I mean, could it be a useful trobuleshooting/operational monitoring aid? Or does this happen frequently and does not indicate a a problem (actually does not look so to me).

Reply:
LogMsg is used in line below inside net_ossl_lastOpenSSLErrorMsg function.